test/jdk/java/lang/ClassLoader/loadLibraryDeadlock/TestLoadLibraryDeadlock.java

```diff
@@ -0,0 +1,237 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8266310
+ * @summary deadlock while loading the JNI code
+ * @library /test/lib
+ * @build LoadLibraryDeadlock Class1 p.Class2
+ * @run main/othervm/native -Xcheck:jni TestLoadLibraryDeadlock
+ */
+
+import jdk.test.lib.Asserts;
+import jdk.test.lib.JDKToolFinder;
+import jdk.test.lib.process.*;
+
+import java.lang.ProcessBuilder;
+import java.lang.Process;
+import java.io.*;
+import java.util.*;
+import java.util.concurrent.*;
+
+public class TestLoadLibraryDeadlock {
+
+    private static final String KEYSTORE = "keystore.jks";
+    private static final String STOREPASS = "changeit";
+    private static final String KEYPASS = "changeit";
+    private static final String ALIAS = "test";
+    private static final String DNAME = "CN=test";
+    private static final String VALIDITY = "366";
+
+    private static String testClassPath = System.getProperty("test.classes");
+    private static String testLibraryPath = System.getProperty("test.nativepath");
+    private static String classPathSeparator = System.getProperty("path.separator");
+
+    private static OutputAnalyzer runCommand(File workingDirectory, String... commands) throws Throwable {
+        ProcessBuilder pb = new ProcessBuilder(commands);
+        pb.directory(workingDirectory);
+        System.out.println("COMMAND: " + String.join(" ", commands));
+        return ProcessTools.executeProcess(pb);
+    }
+
+    private static OutputAnalyzer runCommandInTestClassPath(String... commands) throws Throwable {
+        return runCommand(new File(testClassPath), commands);
+    }
+
+    private static OutputAnalyzer genKey() throws Throwable {
+        runCommandInTestClassPath("rm", "-f", KEYSTORE);
+        String keytool = JDKToolFinder.getJDKTool("keytool");
+        return runCommandInTestClassPath(keytool,
+                "-storepass", STOREPASS,
+                "-keypass", KEYPASS,
+                "-keystore", KEYSTORE,
+                "-keyalg", "rsa", "-keysize", "2048",
+                "-genkeypair",
+                "-alias", ALIAS,
+                "-dname", DNAME,
+                "-validity", VALIDITY
+        );
+    }
+
+    private static OutputAnalyzer createJar(String outputJar, String... classes) throws Throwable {
+        String jar = JDKToolFinder.getJDKTool("jar");
```

Comment:
You can consider using `jdk.test.lib.util.JarUtils` to reduce the test execution time so that it can create the jar without exec'ing another process.

Reply:
I tried, but this doesn't work that well unfortunately with JarUtils.createJar() as it only produces a valid jar-file capable of triggering the issue when the compiled class files (jtreg @build output) are in the same directory as the current directory of the process calling JarUtils.createJar(). Jtreg @build outputs the compiled classes to a separate directory. Creation of a new jar process allows to change the current directory so that the relative paths to the class files are such that it would form the valid output jar-file. Alternatively we could copy the class files to the current directory prior to creating the jar-file, but that would introduce an extra step. Would you mind me keeping the createJar() in the test?

Reply:
Is `jar -C <dir>` option what you want?   FYI.  `jar` tool is a tool provider which is a simple way to run `jar` tool in process (lookup the tool provider by calling `ToolProvider.findFirst("jar")`).

Yes I am fine with what you have if this does not work.